text/0000-lts.md

```diff
@@ -0,0 +1,164 @@
+becomes the new stable.
+
+To keep it as easy as possible to remember which releases are lts release, we
+will begin the lts schedule so that the point version of the LTS release is
+always divisible by four. Depending on implementation schedule, that means that
+the first LTS release will probably be 1.28 or 1.32, release concurrently with
+either 1.29 or 1.33.
+
+## Level of support from the Rust project
+
+All three of the stable release channels - stable, lts, and lts-prior - will
+receive the same level of support. We will host documentation for all three
+channels and provide equivalent levels of patch release support - that is, if a
+bug justifying a patch exists in any of these channels, we will release a patch
+for that channel, just as we do today for stable. However, lts releases are
+less likely to receive patch releases in practice, since these bugs are most
+often found in the first six weeks after a stable release.
+
+As a result of this, a stable release which is a candidate for the lts channel
+(that is, its minor version is a multiple of four) will be supported across all
+three channels for the following nine release cycles, a period of 54 weeks or
+just over a year.
+
+## Recommendations for library authors
+
+Today, library authors fall into three general camps:
+
+1. Their library compiles on the nightly channel.
+2. Their library compiles on the stable channel.
+3. Their library compiles on any release more recent than an arbitrary old
+release, which is different for each library (for example, `failure` compiles
+on 1.18.0, whereas `serde` compiles on 1.13.0).
+
+Frequently, whichever level of support they offer, they guarantee it by testing
+their code on that channel or release using CI.
+
+The Rust project recommends that library authors interested in providing
+support for users on versions older than the latest stable, instead of focusing
+on supporting particular *versions*, focus on supporting particular *release
+channels*. That is, a library may guarantee that it will compile on any of the
+stable release channels, determining how old the version of Rust their users
+can be on with a guarantee of compatibility.
+
+We recommend that libraries consider dropping support for a release channel to
+be a breaking change, but that dropping support for any particular version of
+Rust to be a non-breaking change. 
+
+If a library does not support a particular release channel, effectively all of
+the libraries that depend on *it* also do not support that channel. For this
+reason, we strongly recommend that so-called "core libraries" - those on which
+many other libraries depend - adopt a practice of supporting all three stable
+release channels. Examples of core libraries include libc and serde.
+
+For libraries that do not believe they are "core libraries," we make no
+particular recommendation about which channels to support. There is always a
+trade off between providing the best API for users on stable as soon as
```

Comment:
In principle, I have no objection to providing LTS release channels if the bandwidth exists..

However, I believe the timing is not right and that *this should be done in 2021*. Why? Because the existence of LTS and the marketing thereof incentivizes supporting LTS releases and users will request that of libraries. I think that for the time being, there are important language features, such as:
+ Generic Associated Types (GATs),
+ const generics,
+ support for impl Trait in traits,
+ specialization

that would improve many APIs considerably, but which will not make it into Rust 2018.

Therefore, I think we should wait until some important language features are in place.

Reply:
@Centril But LTS releases, as proposed, are to supported only for about a year, so should not intefer with marketing of these language features. That is, the conservative people would have needed to upgrade 2-3 times anyways to get to 2021.

Reply:
It is not the marketing of the features I'm concerned about, but how LTS releases and marketing thereof incentivizes libraries, to *become*, in my view, prematurely more conservative, and thus miss out on these *specific* features that are important for the public API of libraries. 2021 was just an example; it doesn't have to be that long. But I still think the time is not right, yet.

Reply:
@Centril I think the important libraries are already conservative anyways, so I think they do not bump minimum versions unless they "have to". What you are also asking is for someone to decide what future feature is worth waiting for, meanwhile we were able to get by without such features.

Reply:
@tshepang I don't think that is true. Some important libraries are conservative (`serde`). Other don't compile on stable yet (`rocket`). Other important libraries have breaking changes more frequently (`syn`). A library like `diesel` supports 1.24.1, which was released 2018-03-01, in its `.travis.yml` (cc @sgrif, double check this claim?) Some libraries will/should be important but can't be made 1.0 yet due to lack of the specific features I've cited, in particular GATs.

> What you are also asking is for someone to decide what future feature is worth waiting for, meanwhile we were able to get by without such features.

I think those features in particular, and especially GATs affect very much the fundamental nature of some library APIs (see `StreamingIterator` vs. `Iterator`). One can always *"get by"* with a very minimal mostly featureless languages without parametric polymorphism and such things. But if you want more than "getting by", then good APIs depend on the abstractive power of the language.

I am not asking that everyone should wait for these features. They can always pin to some version of Rust like 1.13 as they've done so far. But if we're going to provide LTS releases, then our story around conditional compilation should first be improved, and these features, in particular GATs, important for the APIs of libraries, should be there first, in my opinion.

Reply:
rocket and syn are special cases, the former explicitly target nightly, and the latter depends on recent compiler changes (I think), and I besides didn't consider it *important*, not in a sense that serde is important. The diesel one falls under "have to", as explained here https://github.com/rust-lang/rfcs/pull/2483#pullrequestreview-131029414.

Reply:
`rocket` targets nightly because it has no choice but to (until proc macros land), which is precisely my point. There are fundamental language features which have not landed yet that important crates need. `diesel`, a central crate in the ecosystem, having to support 1.24.1 means that it wouldn't otherwise. This means that it would be even *less* conservative.